### PR TITLE
Use getBaseDir not base_dir document attribute.

### LINKED ISSUE #62: Path resolution incorrect if no output directory specified
I'm looking at updating the vscode extension to the latest version of `asciidoctor-kroki` as part of https://github.com/asciidoctor/asciidoctor-vscode/issues/301

My working document is (say) here:

`/media/mulhollandd/KINGSTON/operating_information/Operating_Information/architecture/Operating_Information_Architecture.adoc`

and the document sets:

```asciidoc
:imagesdir: media
:kroki-server-url: https://kroki.io
:kroki-fetch-diagram:
```

The extension does set the `base_dir` **API option** but not `imagesoutdir`, `outdir`, `to_dir` or `base_dir` document attributes as appear to be expected in `fetch.js`:

https://github.com/Mogztter/asciidoctor-kroki/blob/f9b9abfb64be2339bde78bb77c41339e88dfcbfd/src/fetch.js#L5-L9

Under these circumstances, my images turn up in `/home/mulhollandd/media/` which surprised me no end and it was only looking through the code which made me look :wink:

The cause is `fs.writeFileSync` in:

https://github.com/Mogztter/asciidoctor-kroki/blob/f9b9abfb64be2339bde78bb77c41339e88dfcbfd/src/node-fs.js#L7-L12

where `filePath` might be something like `media/ba38556d595f9b57735747d9de9ee77e7335cdb2.svg`.

The node `fs` module which uses the `fs.openSync` on this relative path and somewhat deep in node internals it appears that if there is only a relative path then perhaps it uses the user's home directory (or perhaps this is what vs code is using as the cwd, although I doubt it). The [node filesystem docs](https://nodejs.org/api/fs.html#fs_file_paths) don't help me very much. 

I think the problem is that the asciidoctor-kroki extension is incorrectly handling `base_dir` which is an API option, not a document attribute. Perhaps we should be using instead the document attribute `docdir`?

https://asciidoctor.org/docs/user-manual/#piping-content-through-the-cli

Looking at the (Ruby) code:

https://github.com/asciidoctor/asciidoctor/blob/8f838709af08cdd5c3f1c91e4a6bba54e87cca54/lib/asciidoctor/convert.rb#L12-L20

It would seem that if `docdir` is used then it is set base on  `base_dir` if it is specified.

https://github.com/asciidoctor/asciidoctor/blob/88bc891d6a0143f3a2142f41d29cabe3cb21fe3f/lib/asciidoctor/document.rb#L380-L390

So I think the correct fix would be to replace `baseDir` with `docDir` along the lines of:

```js
  const imagesOutputDir = doc.getAttribute('imagesoutdir')
  const outDir = doc.getAttribute('outdir')
  const toDir = doc.getAttribute('to_dir')
  const docDir = doc.getAttribute('docdir') || ''
  const imagesDir = doc.getAttribute('imagesdir') || ''
  let dirPath
  if (imagesOutputDir) {
    dirPath = imagesOutputDir
  } else if (outDir) {
    dirPath = path.join(outDir, imagesDir)
  } else if (toDir) {
    dirPath = path.join(toDir, imagesDir)
  } else {
    dirPath = path.join(docDir, imagesDir)
  }
```

but I haven't entirely convinced myself so I haven't submitted a PR.

Thoughts?

### PR DESCRIPTION
**Pull request by danyill**
Closes #62.

### CODE REVIEW

**Comment by Mogztter** [comment]
@danyill So `getBaseDir` actually fixes this issue?

File: test/delDir.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Until recursive: true is a stable part of Node
```
Comment (Mogztter) [inline test/delDir.js:1]: It hurts every time you need to remove recursively in Node :disappointed:

File: test/test.js [hunk 2]
```diff
@@ -83,6 +85,24 @@ plantuml::test/fixtures/alice.puml[svg,role=sequence]
...
     })
+    it('should create diagrams in imagesdir if kroki-fetch-diagram is set', function (done) {
+      const registry = asciidoctor.Extensions.create()
+      asciidoctorKroki.register(registry)
+      const file = `${__dirname}/fixtures/fetch/doc.adoc`
+      const doc = asciidoctor.convertFile(file, { extension_registry: registry, safe: 'unsafe' })
+      fs.unlinkSync(doc.getAttributes().outfile)
+      const imageLocation = path.join(doc.base_dir, doc.getAttributes().imagesdir)
+      fs.readdir(imageLocation, (err, files) => {
+        if (err) {
```
Comment (Mogztter) [inline test/test.js:96]: Are you sure that an error is thrown if the directory does not exist? I'm not familiar with this API maybe it returns an empty list?

File: test/test.js [hunk 3]
```diff
@@ -83,6 +85,24 @@ plantuml::test/fixtures/alice.puml[svg,role=sequence]
...
       expect(html).to.contain(`<img src=".asciidoctor/kroki/diag-${hash}.svg" alt="diagram">`)
     })
+    it('should create diagrams in imagesdir if kroki-fetch-diagram is set', function (done) {
+      const registry = asciidoctor.Extensions.create()
+      asciidoctorKroki.register(registry)
+      const file = `${__dirname}/fixtures/fetch/doc.adoc`
+      const doc = asciidoctor.convertFile(file, { extension_registry: registry, safe: 'unsafe' })
+      fs.unlinkSync(doc.getAttributes().outfile)
+      const imageLocation = path.join(doc.base_dir, doc.getAttributes().imagesdir)
+      fs.readdir(imageLocation, (err, files) => {
```
Comment (Mogztter) [inline test/test.js:95]: I think we should use `fsPromises.readdir` https://nodejs.org/dist/latest-v10.x/docs/api/fs.html#fs_fspromises_readdir_path_options

```suggestion
      const files = await fsPromises.readdir(imageLocation)
```

File: test/test.js [hunk 4]
```diff
@@ -83,6 +85,24 @@ plantuml::test/fixtures/alice.puml[svg,role=sequence]
       const hash = rusha.createHash().update(`https://kroki.io/plantuml/svg/${encode(file)}`).digest('hex')
       expect(html).to.contain(`<img src=".asciidoctor/kroki/diag-${hash}.svg" alt="diagram">`)
     })
+    it('should create diagrams in imagesdir if kroki-fetch-diagram is set', function (done) {
```
Comment (Mogztter) [inline test/test.js:88]:
```suggestion
    it('should create diagrams in imagesdir if kroki-fetch-diagram is set', async () => {
```

File: test/test.js [hunk 5]
```diff
@@ -83,6 +85,24 @@ plantuml::test/fixtures/alice.puml[svg,role=sequence]
...
+      asciidoctorKroki.register(registry)
+      const file = `${__dirname}/fixtures/fetch/doc.adoc`
+      const doc = asciidoctor.convertFile(file, { extension_registry: registry, safe: 'unsafe' })
+      fs.unlinkSync(doc.getAttributes().outfile)
+      const imageLocation = path.join(doc.base_dir, doc.getAttributes().imagesdir)
+      fs.readdir(imageLocation, (err, files) => {
+        if (err) {
+          done(`Image location: ${imageLocation} does not exist`)
+        }
+        if (files.length === 1) {
```
Comment (Mogztter) [inline test/test.js:99]: This should be an assertion:

```suggestion
        expect(files.length).to.equal(1)
```

Comment (Mogztter) [reply test/test.js:99]: We can also use:

```suggestion
        expect(files).to.have.lengthOf(1)
```

File: test/test.js [hunk 6]
```diff
@@ -83,6 +85,24 @@ plantuml::test/fixtures/alice.puml[svg,role=sequence]
...
+          // tidy up
+          delDir.deleteDirWithFiles(imageLocation)
```
Comment (Mogztter) [inline test/test.js:101]: I think we should do it in a `finally` block

File: test/test.js (hunk 2 above)
Comment (danyill) [reply test/test.js:96]: It did, but I have changed as you recommended below.

File: test/test.js (hunk 6 above)
Comment (danyill) [reply test/test.js]: I couldn't quite figure out how to do this.

**Comment by danyill** [comment]
@Mogztter I've tried to follow your suggestions, just the last one wasn't apparent to me. I've rebased to master.
Yes, `getBaseDir()` does fix this issue.

File: test/test.js [hunk 7]
```diff
@@ -83,7 +86,19 @@ plantuml::test/fixtures/alice.puml[svg,role=sequence]
...
+      const registry = asciidoctor.Extensions.create()
+      asciidoctorKroki.register(registry)
+      const file = `${__dirname}/fixtures/fetch/doc.adoc`
+      const doc = asciidoctor.convertFile(file, { extension_registry: registry, safe: 'unsafe' })
+      fs.unlinkSync(doc.getAttributes().outfile)
+      const imageLocation = path.join(doc.base_dir, doc.getAttributes().imagesdir)
+      await fsPromises.readdir(imageLocation).then((files) => {
+        expect(files).to.have.lengthOf(1)
+        delDir.deleteDirWithFiles(imageLocation)
+      })
```
Comment (Mogztter) [inline test/test.js:99]:
```suggestion
      try {
        const files = await fsPromises.readdir(imageLocation)
        expect(files).to.have.lengthOf(1)
      } finally {
        delDir.deleteDirWithFiles(imageLocation)
      }
```

File: test/test.js (hunk 6 above)
Comment (Mogztter) [reply test/test.js]: That's OK, I've added a commit if you want to have a look: https://github.com/Mogztter/asciidoctor-kroki/pull/64/commits/9325af6c6190bc3511c9c97a10cee4fbbbf3fb2f

**Comment by Mogztter** [comment]
>  I've tried to follow your suggestions, just the last one wasn't apparent to me. I've rebased to master.

Thanks, the code is looking good :+1: 
I've added the try/finally block so we are good to go!

> Yes, getBaseDir() does fix this issue.

Perfect :ok_hand:

**Comment by danyill** [comment]
Thanks for the support and assistance :+1:

**Pull request merged**